ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/ReadJobImpl.java

```diff
@@ -156,6 +159,103 @@ public void transferItem(final Ds3Client client, final BulkObject ds3Object) thr
         }
     }
 
+    private final class GetObjectTransferrerNetworkFailureDecorator implements ItemTransferrer {
+        private final JobState jobState;
+        private Long numBytesToTransfer;
+        private ImmutableCollection<Range> ranges;
+        private final AtomicLong destinationChannelOffset = new AtomicLong(0);
+        private final AtomicReference<ItemTransferrer> itemTransferrer = new AtomicReference<>();
+
+        private GetObjectTransferrerNetworkFailureDecorator(final JobState jobState) {
+            this.jobState = jobState;
+            itemTransferrer.set(new GetObjectTransferrer(jobState));
+        }
+
+        @Override
+        public void transferItem(final Ds3Client client, final BulkObject ds3Object) throws IOException {
+            try {
+                itemTransferrer.get().transferItem(client, ds3Object);
+            } catch (final ContentLengthNotMatchException contentLengthNotMatchException) {
+                makeNewItemTransferrer(ds3Object, contentLengthNotMatchException.getTotalBytes());
+
+                emitContentLengthMismatchFailureEvent(ds3Object, contentLengthNotMatchException);
+
+                throw new RecoverableIOException(contentLengthNotMatchException);
+            }
+        }
+
+        private void makeNewItemTransferrer(final BulkObject ds3Object, final long numBytesTransferred) {
+            initializeRangesAndTransferSize(ds3Object);
+            updateRanges(numBytesTransferred);
+            destinationChannelOffset.getAndAdd(numBytesTransferred);
+
+            itemTransferrer.set(new GetPartialObjectTransferrer(jobState, ranges, destinationChannelOffset.get()));
+        }
+
+        private synchronized void initializeRangesAndTransferSize(final BulkObject ds3Object) {
+            if (ranges == null) {
+                ranges = getRangesForBlob(blobToRanges, ds3Object);
+            }
+
+            if (ranges == null) {
+                final long numBytesTransferred = 0;
+                ranges = RangeHelper.replaceRange(ranges, numBytesTransferred, ds3Object.getLength());
+            }
+
+            if (numBytesToTransfer == null) {
+                numBytesToTransfer = RangeHelper.transferSizeForRanges(ranges);
+            }
+        }
+
+        private synchronized void updateRanges(final long numBytesTransferred) {
+            ranges = RangeHelper.replaceRange(ranges, numBytesTransferred, numBytesToTransfer);
+        }
+
+        private void emitContentLengthMismatchFailureEvent(final BulkObject ds3Object, final Throwable t) {
+            final FailureEvent failureEvent = new FailureEvent.Builder()
```

Comment:
I tend to favor the syntax that guava uses for its builders.  ie `FailureEvent.builder()` rather than `new FailureEvent.Builder()`